src/theory/strings/regexp_elim.cpp

```diff
@@ -182,9 +184,29 @@ Node RegExpElimination::eliminateConcat(Node atom)
         // would have been the case than "ABB" would be a model for x, where
         // the second constraint refers to the third position, and the third
         // constraint refers to the second position.
-        Node fit = nm->mkNode(gap_exact[sep_children.size() - 1] ? EQUAL : LEQ,
-                              nm->mkNode(MINUS, prev_end, lenSc),
-                              loc);
+        Node fit;
+        if( !gap_exact[sep_children.size() - 1] )
+        {
+          // With respect to the above example, this is an optimization. For
+          // that example, we instead produce:
+          //     x in (re.++ "A" _ (re.* _) "B" _) --->
+          //       substr( x, 0, 1 ) = "A" ^          // find "A"
+          //       substr( x, len(x)-2, 1 ) = "B" ^   // "B" is at end - 2
+          //       2 <= len( x ) - 2
+          // The intuition is that above, there are two constraints that insist
+          // that "B" is found, whereas we only need one. The last constraint
+          // above says that the "B" we find at end-2 can be found >=1 after
+          // the "A". Notice we only need two constraints that talk about the
+          // location of finding "B" if both the last gap and second-to-last
+          // gaps are exact.
```

Comment:
I think that we can generalize this optimization to turn all the exact gaps at the end of the regexp into `substr`s that are relative to the end of the string. The way this could work:

1. Iterate over the regexp from the beginning and generate `substr` constraints as long as the gaps are exact.
2. Iterate over the regexp from the end and generate `substr` constraints as along as the gaps are exact.
3. Iterate over the rest of the regexp and generate constraints relative to `substr(x, i, j - i)` where `i` is the last position of an exact constraint and `j` is the (symbolic) first position of the exact constraints from the end
4. Generate a length constraint